virttest/utils_logfile.py

```diff
@@ -0,0 +1,131 @@
+"""
+Control log file utility functions.
+An easy way to log lines to files when the logging system can't be used
+
+:copyright: 2020 Red Hat Inc.
+"""
+import os
+import time
+import threading
+
+from avocado.utils import path as utils_path
+from aexpect.utils.genio import _open_log_files
+from avocado.utils.astring import string_safe_encode
+
+from virttest import data_dir
+
+
+_log_file_dir = data_dir.get_tmp_dir()
+_log_lock = threading.RLock()
+
+
+def _acquire_lock(lock, timeout=10):
+    """
+    Check if the lock is available
+
+    :param lock: threading.RLock object
+    :param timeout: time to Wait for the lock
+
+    :return: boolean. True if the lock is available
+                      False if the lock is unavailable
+    """
+    end_time = time.time() + timeout
+    while time.time() < end_time:
+        if lock.acquire(False):
+            return True
+        time.sleep(0.05)
+    return False
+
+
+class LogLockError(Exception):
+    pass
+
+
+def log_line(filename, line):
+    """
+    Write a line to a file.
+
+    :param filename: Path of file to write to, either absolute or relative to
+                     the dir set by set_log_file_dir().
+    :param line: Line to write.
+    :raise LogLockError: If the lock is unavailable
+    """
+    global _open_log_files, _log_file_dir, _log_lock
+
+    if not _acquire_lock(_log_lock):
+        raise LogLockError("Could not acquire exclusive lock to access"
+                           " _open_log_files")
+    log_file = get_log_filename(filename)
+    base_file = os.path.basename(log_file)
+    try:
+        if base_file not in _open_log_files:
+            # First, let's close the log files opened in old directories
+            close_log_file(base_file)
```

Comment:
just a little wonder  that in line 55, it call require_lock once, but in close_log_file, it call require_lock again, not sure whether they are conflict here.

Reply:
I think it should be the reason that 'f.close()' need a lock to operate.